src/logic/safe/store/actions/createTransaction.ts

```diff
@@ -78,181 +75,253 @@ const navigateToTx = (safeAddress: string, txDetails: TransactionDetails) => {
   history.push(txRoute)
 }
 
-export const createTransaction =
-  (
-    {
-      safeAddress,
-      to,
-      valueInWei,
-      txData = EMPTY_DATA,
-      notifiedTransaction,
-      txNonce,
-      operation = Operation.CALL,
-      navigateToTransactionsTab = true,
-      origin = null,
-      safeTxGas: safeTxGasArg,
-      ethParameters,
-      delayExecution = false,
-    }: CreateTransactionArgs,
-    onUserConfirm?: ConfirmEventHandler,
-    onError?: ErrorEventHandler,
-  ): CreateTransactionAction =>
-  async (dispatch: Dispatch, getState: () => AppReduxState): Promise<DispatchReturn> => {
-    const state = getState()
+const getSafeTxGas = async (txProps: RequiredTxProps, safeVersion: string): Promise<string> => {
+  const estimationProps: SafeTxGasEstimationProps = {
+    safeAddress: txProps.safeAddress,
+    txData: txProps.txData,
+    txRecipient: txProps.to,
+    txAmount: txProps.valueInWei,
+    operation: txProps.operation,
+  }
 
-    const ready = await onboardUser()
-    if (!ready) return
+  let safeTxGas = '0'
+  try {
+    safeTxGas = await estimateSafeTxGas(estimationProps, safeVersion)
+  } catch (err) {
+    logError(Errors._617, err.message)
+  }
+  return safeTxGas
+}
 
-    const { account: from, hardwareWallet, smartContractWallet } = providerSelector(state)
-    const safeVersion = currentSafeCurrentVersion(state) as string
-    const safeInstance = getGnosisSafeInstanceAt(safeAddress, safeVersion)
-    const chainId = currentChainId(state)
+export class TxSender {
+  notifications: ReturnType<typeof createTxNotifications>
+  nonce: string
+  isExecution: boolean
+  txArgs: TxArgs
+  safeTxHash: string
+  txProps: RequiredTxProps
+  from: string
+  dispatch: Dispatch
+  safeVersion: string
+
+  // On transaction completion (either confirming or executing)
+  async onComplete(signature?: string, confirmCallback?: ConfirmEventHandler): Promise<void> {
+    const { txArgs, safeTxHash, txProps, dispatch, notifications } = this
+
+    let txDetails: TransactionDetails | null = null
+
+    // Send the tx to the backend if
+    // 1) It's a confirmation w/o exection
+    // 2) It's a creation + execution w/o pre-approved signatures
+    if (!this.isExecution || !signature) {
+      try {
+        txDetails = await saveTxToHistory({ ...txArgs, signature, origin })
+      } catch (err) {
+        // catch
+        return
+      }
+    }
 
-    const lastTx = getLastTransaction(state)
-    let nextNonce: string
-    try {
-      nextNonce = (await getRecommendedNonce(safeAddress)).toString()
-    } catch (e) {
-      logError(Errors._616, e.message)
-      nextNonce = await safeInstance.methods.nonce().call()
+    notifications.closePending()
+
+    // This is used to communicate the safeTxHash to a Safe App caller
+    confirmCallback?.(safeTxHash)
+
+    // Go to a tx deep-link
+    if (txDetails && txProps.navigateToTransactionsTab) {
+      navigateToTx(txProps.safeAddress, txDetails)
```

Comment:
If we are navigating to a deeplink, we should add a `return`. We don't need to fetch the transactions again (two lines below).